### PR TITLE
hash: Add option to set custom hash function

### PR DESCRIPTION
**Pull request by BenjD90**
Hello,

While using mingo, I faced on hash collision that are blocking some usages.

Here a PR to allow users to give there own hash function if the prefer to.

I added a test with the collision values I got, I named it `hash function collision`.

@kofrasa Could you review it ? It is related to the issue #62 

Thanks in advance :smiley:

### CODE REVIEW

File: src/core.ts [hunk 1]
```diff
@@ -25,12 +25,18 @@ export interface CollationSpec {
   readonly backwards?: boolean // unsupported
 }
 
+/**
+ * Custom function to hash values to improve faster comparaisons
+ */
+export type CustomHashFunction =  (v: string) => string | number
```
Comment (kofrasa) [inline src/core.ts:31]: **recommendation:** could we simplify the type name to `HashFunction`

File: src/core.ts [hunk 2]
```diff
@@ -25,12 +25,18 @@ export interface CollationSpec {
...
+
 /**
  * Generic options interface passed down to all operators
  */
 export interface Options {
   readonly idKey: string;
-  readonly collation?: CollationSpec
+  readonly collation?: CollationSpec;
+  readonly customHashFunction?:CustomHashFunction
```
Comment (kofrasa) [inline src/core.ts:39]: similar to above, the field could be `hashFunction`

File: src/util.ts [hunk 3]
```diff
@@ -220,9 +221,10 @@ export function merge(target: object, obj: object, options?: MergeOptions): obje
  *
  * @param  {Array} xs The first array
  * @param  {Array} ys The second array
+ * @param  {Function} customHashFunction Custom function to hash values, default the hashCode method
  * @return {Array}    Result array
  */
-export function intersection(a: any[], b: any[]): any[] {
+export function intersection(a: any[], b: any[], customHashFunction: CustomHashFunction): any[] {
```
Comment (kofrasa) [inline src/util.ts:227]: `hashFunction` should be optional. same for the other affected functions; `union`, `memoize`, and `groupBy`.

File: src/util.ts [hunk 4]
```diff
@@ -409,11 +411,15 @@ function encode(value: any): string {
...
+export function hashCode(value: any, customHashFunction?: CustomHashFunction): number | string | null {
   if (isNil(value)) return null
 
   let hash = 0
   let s = encode(value)
+	if(customHashFunction) {
+		return customHashFunction(value)
+	}
+
   let i = s.length
```
Comment (kofrasa) [inline src/util.ts:423]: **recommendation:** I suggest moving the default hashing algorithm into a constant say `DEFAULT_HASH_FUNC` and using that as a default argument on the `hashCode` function.

```typescript
const DEFAULT_HASH_FUNC = ...

export function hashCode(value: any, hashFunction: HashFunction = DEFAULT_HASH_FUNC)
```

**Review by kofrasa — CHANGES_REQUESTED** [review]
Left some minor comments. Thanks for this PR and for adding unit tests.

File: src/util.ts (hunk 4 above)
Comment (BenjD90) [reply src/util.ts:423]: done I've change the HashFunction signature to include the encoding part in the hash function

File: src/util.ts (hunk 3 above)
Comment (BenjD90) [reply src/util.ts]: done

File: src/core.ts (hunk 2 above)
Comment (BenjD90) [reply src/core.ts]: renamed to hashFunction with type HashFunction

File: src/core.ts (hunk 1 above)
Comment (BenjD90) [reply src/core.ts]: done, as you said

**Comment by BenjD90** [comment]
@kofrasa I've applied your recommendations, I let you resolve the conversations if it's ok for you :)

**Comment by BenjD90** [comment]
@kofrasa Thanks for the merge 🎉 Looking forward to see it in a release :smile:

**Pull request merged**